### PR TITLE
code: improve expr_rewriter test coverage

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <chris@lonng.org>

### What problem does this PR solve? 

Improve test coverage.

### What is changed and how it works?

Improve test coverage.

### Check List 

Tests 

 - Unit test

Side effects

N/A

Related changes

N/A

### CODE REVIEW

**Comment by lonng** [comment]
PTAL @kennytm @amyangfei

File: .codecov.yml [hunk 1]
```diff
@@ -33,6 +33,7 @@ ignore:
   - "LICENSES"
   - "*_test.go"
   - "marker.go" # This file only contains empty function stub
+  - "http.go" # Ignore http request handler
```
Comment (lonng) [inline .codecov.yml]:
```suggestion
  - "http.go" # Ignore the http request handler
```

File: .codecov.yml [hunk 2]
```diff
@@ -33,6 +33,7 @@ ignore:
   - "LICENSES"
   - "*_test.go"
   - "marker.go" # This file only contains empty function stub
+  - "http.go" # Ignore the http request handler
```
Comment (kennytm) [inline .codecov.yml]: Unlike `marker.go` the `http.go` does contain logic and I think it should be tested eventually (by writing an `http_test.go`). Please remove this line.

File: code/rewriter_test.go [hunk 3]
```diff
@@ -1464,44 +1512,30 @@ func unittest() {
...
-	err := os.MkdirAll(s.path, os.ModePerm)
-	c.Assert(err, IsNil)
-	for _, cs := range cases {
-		original := filepath.Join(s.path, cs.filepath)
-		err := ioutil.WriteFile(original, []byte(cs.original), os.ModePerm)
-		c.Assert(err, IsNil)
-	}
+		{
+			filepath: "bad-basic-test2.go",
+			errormsg: `failpoint.Inject: closure signature illegal .*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: closure signature illegal .*`,
```

File: code/rewriter_test.go [hunk 4]
```diff
@@ -1464,44 +1512,30 @@ func unittest() {
...
-func (s *rewriterSuite) TestRewriteBad2(c *C) {
-	var cases = []struct {
-		filepath string
-		original string
-	}{
-		// bad cases
 		{
-			filepath: "bad-basic-test2.go",
+			filepath: "bad-basic-test3.go",
+			errormsg: `failpoint.Inject: closure signature illegal .*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: closure signature illegal .*`,
```

File: code/rewriter_test.go [hunk 5]
```diff
@@ -1512,50 +1546,36 @@ import (
...
-	err := os.MkdirAll(s.path, os.ModePerm)
-	c.Assert(err, IsNil)
-	for _, cs := range cases {
-		original := filepath.Join(s.path, cs.filepath)
-		err := ioutil.WriteFile(original, []byte(cs.original), os.ModePerm)
-		c.Assert(err, IsNil)
-	}
+		{
+			filepath: "bad-basic-test4.go",
+			errormsg: `failpoint.Inject: expect 2 arguments but got 3`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: expect 2 arguments but got 3`,
```

File: code/rewriter_test.go [hunk 6]
```diff
@@ -1512,50 +1546,36 @@ import (
...
-func (s *rewriterSuite) TestRewriteBad3(c *C) {
-	var cases = []struct {
-		filepath string
-		original string
-	}{
-		// bad cases
 		{
-			filepath: "bad-basic-test3.go",
+			filepath: "bad-basic-test5.go",
+			errormsg: `failpoint.Inject: closure signature illegal .*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: closure signature illegal .*`,
```

File: code/rewriter_test.go [hunk 7]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+	failpoint.Inject("failpoint-name", func(val, val2 int) {
+		fmt.Println("unit-test", val)
+	})
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-test6.go",
+			errormsg: `failpoint.Inject: first argument expect string literal but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: first argument expect string literal but got.*`,
```

File: code/rewriter_test.go [hunk 8]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+	failpoint.Inject(func(val int, val2 int) {
+		fmt.Println("unit-test", val)
+	}, "failpoint-name")
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-test7.go",
+			errormsg: `failpoint.Inject: second argument expect closure but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Inject: second argument expect closure but got.*`,
```

File: code/rewriter_test.go [hunk 9]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.Inject("failpoint-name", "failpoint-name")
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test1.go",
+			errormsg: `failpoint.InjectContext: expect 3 arguments but got 4`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: expect 3 arguments but got 4`,
```

File: code/rewriter_test.go [hunk 10]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+	failpoint.InjectContext("failpoint-name", nil, nil, func(val int, val2 int) {
+		fmt.Println("unit-test", val)
+	})
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test2.go",
+			errormsg: `failpoint.InjectContext: first argument expect context but go.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: first argument expect context but got.*`,
```

File: code/rewriter_test.go [hunk 11]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+	failpoint.InjectContext(func(){}, nil, func(val int, val2 int) {
+		fmt.Println("unit-test", val)
+	})
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test3.go",
+			errormsg: `failpoint.InjectContext: second argument expect string literal but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: second argument expect string literal but got.*`,
```

File: code/rewriter_test.go [hunk 12]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+	failpoint.InjectContext(nil, func(){}, func(val int, val2 int) {
+		fmt.Println("unit-test", val)
+	})
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test4.go",
+			errormsg: `failpoint.InjectContext: third argument expect closure but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: third argument expect closure but got.*`,
```

File: code/rewriter_test.go [hunk 13]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.InjectContext(nil, "failpoint-name", "string literal")
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test5.go",
+			errormsg: `failpoint.InjectContext: closure signature illegal.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: closure signature illegal.*`,
```

File: code/rewriter_test.go [hunk 14]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.InjectContext(nil, "failpoint-name", func(val int, val int) {})
+}
+`,
+		},
+
+		{
+			filepath: "bad-basic-ctx-test6.go",
+			errormsg: `failpoint.InjectContext: closure signature illegal.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.InjectContext: closure signature illegal.*`,
```

File: code/rewriter_test.go [hunk 15]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.InjectContext(nil, "failpoint-name", func(val, val int) {})
+}
+`,
+		},
+
+		{
+			filepath: "bad-case-break.go",
+			errormsg: `failpoint.Break expect 1 or 0 arguments, but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Break expect 1 or 0 arguments, but got.*`,
```

File: code/rewriter_test.go [hunk 16]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.Break("11", "22")
+}
+`,
+		},
+
+		{
+			filepath: "bad-case-continue.go",
+			errormsg: `failpoint.Continue expect 1 or 0 arguments, but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Continue expect 1 or 0 arguments, but got.*`,
```

File: code/rewriter_test.go [hunk 17]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.Continue("11", "22")
+}
+`,
+		},
+
+		{
+			filepath: "bad-case-label.go",
+			errormsg: `failpoint.Label expect 1 arguments, but got.*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Label expect 1 arguments, but got.*`,
```

File: code/rewriter_test.go [hunk 18]
```diff
@@ -1566,10 +1586,254 @@ import (
...
+
+func unittest() {
+	failpoint.Label("11", "22")
+}
+`,
+		},
+
+		{
+			filepath: "bad-case-goto.go",
+			errormsg: `failpoint.Goto expect 1 arguments, but got .*`,
```
Comment (kennytm) [inline code/rewriter_test.go]:
```suggestion
			errormsg: `failpoint\.Goto expect 1 arguments, but got .*`,
```

File: runtime_test.go [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+var _ = Suite(&runtimeSuite{})
+
+type runtimeSuite struct{}
+
+func (s *runtimeSuite) TestRuntime(c *C) {
+	err := failpoint.Enable("runtime-test-1", "return(1)")
+	c.Assert(err, IsNil)
+	ok, val := failpoint.Eval("runtime-test-1")
+	c.Assert(ok, Equals, true)
```
Comment (kennytm) [inline runtime_test.go]:
```suggestion
	c.Assert(ok, IsTrue)
```

File: runtime_test.go [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+	c.Assert(val.(int), Equals, 1)
+
+	err = failpoint.Enable("runtime-test-2", "invalid")
+	c.Assert(err, ErrorMatches, `failpoint: could not parse terms`)
+
+	ok, val = failpoint.Eval("runtime-test-2")
+	c.Assert(ok, Equals, false)
```
Comment (kennytm) [inline runtime_test.go]:
```suggestion
	c.Assert(ok, IsFalse)
```

File: runtime_test.go [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+
+	err = failpoint.Disable("runtime-test-1")
+	c.Assert(err, IsNil)
+
+	ok, val = failpoint.Eval("runtime-test-1")
+	c.Assert(ok, Equals, false)
```
Comment (kennytm) [inline runtime_test.go]:
```suggestion
	c.Assert(ok, IsFalse)
```

File: runtime_test.go [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+		time.Sleep(time.Second)
+		err := failpoint.Disable("gofail/testPause")
+		c.Assert(err, IsNil)
+		close(ch)
+	}()
+	err = failpoint.Enable("gofail/testPause", "pause")
+	c.Assert(err, IsNil)
+	start := time.Now()
+	ok, v := failpoint.Eval("gofail/testPause")
+	c.Assert(ok, Equals, false)
```
Comment (kennytm) [inline runtime_test.go]:
```suggestion
	c.Assert(ok, IsFalse)
```

File: runtime_test.go [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+	c.Assert(v, IsNil)
+	if time.Since(start) < 100*time.Millisecond {
+		c.Fatal("not paused")
+	}
```
Comment (kennytm) [inline runtime_test.go]:
```go
c.Assert(time.Since(start), GreaterEqual, 100*time.Millisecond, "not paused")
```

File: runtime_test.go [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+
+	err = failpoint.Enable("runtime-test-5", "50*return(5)")
+	c.Assert(err, IsNil)
+	for i := 0; i < 50; i++ {
+		ok, val = failpoint.Eval("runtime-test-5")
+		c.Assert(ok, Equals, true)
+		c.Assert(val.(int), Equals, 5)
+	}
+	ok, val = failpoint.Eval("runtime-test-5")
+	c.Assert(ok, Equals, false)
```
Comment (kennytm) [inline runtime_test.go]:
```suggestion
	c.Assert(ok, IsFalse)
```

File: http_test.go [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+	c.Assert(res.Code, Equals, http.StatusBadRequest)
+	c.Assert(res.Body.String(), Matches, "failed to set failpoint")
+
+	// GET
+	req, err = http.NewRequest(http.MethodGet, "http://127.0.0.1/failpoint-name", strings.NewReader(""))
+	c.Assert(err, IsNil)
+	res = httptest.NewRecorder()
+	handler.ServeHTTP(res, req)
+	c.Assert(res.Code, Equals, http.StatusOK)
+	c.Assert(res.Body.String(), Matches, "return(1)")
```
Comment (kennytm) [inline http_test.go]:
```suggestion
	c.Assert(res.Body.String(), Matches, `return\(1\)`)
```

File: http_test.go [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+	handler.ServeHTTP(res, req)
+	c.Assert(res.Code, Equals, http.StatusBadRequest)
+	c.Assert(res.Body.String(), Matches, "failed to GET")
+
+	req, err = http.NewRequest(http.MethodGet, "http://127.0.0.1/", strings.NewReader(""))
+	c.Assert(err, IsNil)
+	res = httptest.NewRecorder()
+	handler.ServeHTTP(res, req)
+	c.Assert(res.Code, Equals, http.StatusOK)
+	c.Assert(res.Body.String(), Matches, "failpoint-name=return(1)")
```
Comment (kennytm) [inline http_test.go]:
```suggestion
	c.Assert(res.Body.String(), Matches, `failpoint-name=return\(1\)`)
```

**Comment by lonng** [comment]
Ping @amyangfei

**Pull request merged**